plugins/outputs/otlp/client.go

```diff
@@ -0,0 +1,314 @@
+				RootCAs:    certPool,
+			}
+		}
+		_ = level.Debug(c.logger).Log(
+			"msg", "TLS configured",
+			"server", c.url.Hostname(),
+			"root_certs", fmt.Sprint(c.rootCertificates),
+		)
+		dopts = append(dopts, grpc.WithTransportCredentials(credentials.NewTLS(&tcfg)))
+	} else {
+		dopts = append(dopts, grpc.WithInsecure())
+	}
+	if c.compressor != "" && c.compressor != "none" {
+		dopts = append(dopts, grpc.WithDefaultCallOptions(grpc.UseCompressor(c.compressor)))
+	}
+	address := c.url.Hostname()
+	if len(c.url.Port()) > 0 {
+		address = net.JoinHostPort(address, c.url.Port())
+	}
+	conn, err := grpc.DialContext(ctx, address, dopts...)
+	c.conn = conn
+	if err != nil {
+		_ = level.Debug(c.logger).Log(
+			"msg", "connection status",
+			"address", address,
+			"err", err,
+		)
+		return nil, err
+	}
+
+	return conn, err
+}
+
+// Selftest sends an empty request the endpoint.
+func (c *Client) Selftest(ctx context.Context) error {
+	// Loop until the context is canceled, allowing for retryable failures.
+	for {
+		conn, err := c.getConnection(ctx)
+
+		if err == nil {
+			service := metricsService.NewMetricsServiceClient(conn)
+			empty := &metricsService.ExportMetricsServiceRequest{}
+
+			_, err = service.Export(c.grpcMetadata(ctx), empty)
+			if err == nil {
+				return nil
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			if isRecoverable(err) {
+				_ = level.Info(c.logger).Log("msg", "selftest recoverable error, still trying", "err", err)
+				continue
+			}
+		}
+		return fmt.Errorf(
+			"non-recoverable failure in selftest: %s",
+			truncateErrorString(err),
+		)
+	}
+}
+
+// Store sends a batch of samples to the endpoint.
+func (c *Client) Store(req *metricsService.ExportMetricsServiceRequest) error {
+	tss := req.ResourceMetrics
+	if len(tss) == 0 {
+		// Nothing to do, return silently.
+		return nil
+	}
+
+	// Note the call to getConnection() applies its own timeout for Dial().
+	ctx := context.Background()
+	conn, err := c.getConnection(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	service := metricsService.NewMetricsServiceClient(conn)
+
+	errs := make(chan error, len(tss)/maxTimeseriesPerRequest+1)
+	var wg sync.WaitGroup
+	for i := 0; i < len(tss); i += maxTimeseriesPerRequest {
+		end := i + maxTimeseriesPerRequest
+		if end > len(tss) {
+			end = len(tss)
+		}
+		wg.Add(1)
+		go func(begin int, end int) {
+			defer wg.Done()
+			reqCopy := &metricsService.ExportMetricsServiceRequest{
+				ResourceMetrics: req.ResourceMetrics[begin:end],
+			}
+
+			var md metadata.MD
+			var err error
+
+			if _, err = service.Export(c.grpcMetadata(ctx), reqCopy, grpc.Trailer(&md)); err != nil {
+				_ = level.Error(c.logger).Log(
+					"msg", "export failure",
+					"err", truncateErrorString(err),
+					"size", proto.Size(reqCopy),
+					"trailers", fmt.Sprint(md),
+					"recoverable", isRecoverable(err),
+				)
```

Comment:
Please remove the extensive logging on this level. You might want some brief debug messages but errors will be logged anyway.